### ISSUE TITLE
LaTeX preamble embedded for portability?

### ISSUE THREAD

**Issue by pacuserra** [textext/textext#99]
When sharing Inkscape documents containing textext groups, the LaTeX preambles are missing, so other users editing such groups with their textext plugins do not get the same LaTeX result.

It would be nice to have the possibility of embedding the LaTeX preamble code on each textext group.

Maybe a straightforward way would be allowing the use of \begin{document} inside the textext code dialogue. Anything before \begin{document} would be then added to the preamble. In case of \begin{document} not being present, then all the code would be part of the document body, so retro-compatibility with previous textext groups could be preserved.

BTW, thanks a lot for the fantastic textext tool!

**Comment by jcwinkler** [comment]
We thought about the idea to include the used packages in the textext node but then abandoned it because it triggers a lot of other questions: Which packages to use in new nodes, how to edit the used packages, etc. 

Nevertheless, I fully agree that portability from machine to machine is currently a problem if preamble files are placed in different directories. 

Of course, extension made a lot of progress since then, so we can rethink this problem. 

One idea: Offer an edit field, which can hold a comma seperated list of packages, e.g.
```
siunitx, [utf-8]inputenc
```
this could more or less easily be parsed to
```
\usepakage{siunitx}
\usepackage[utf-8]{inputenc}
```
, added to the file to be compiled and stored in the node.

**Comment by sizmailov** [comment]
@pacuserra Good call! 

I had in mind adding extra preamble parameters such as "preamble path relative to svg", "preamble absolute path". It would be similar to what inkscape does for image links. 

Speaking about embedded preamble. It's not a bad idea, but we need to address several issues to handle it properly. Node may contain full preamble text used to generate it.  

@jcwinkler preamble is not limited to packages, it may contain arbitrary stuff, so we need to store it as is.

Say we have "preamble_text" parameter embedded in nodes. 
New node creation goes as before, i.e. reading reading some file from disk. 

If node gets edited we have a choice 
1) use preamble from file
2) use embedded code

It would require change of UI. UI should not grow too much, since most of users are satisfied (i guess) with minimal interface. And we need to support Tk as well (sigh). 

Also we should provide option to edit embedded preamble before node generation. 

From that point there would be two kinds of nodes: with and without embedded preamble. Backward/forward compatibility issues should be carefully analyzed. 

The change should not affect people who does not use this feature. 

It's all doable, but takes time. I may get to it once we finish with current tasks.

**Comment by pacuserra** [comment]
@jcwinkler Sorry, I did not mean to replace the Preamble File but combining it with my proposal. The GUI would only require an "Import" button besides the existing "Clear" one, so the usage could be as follows:

1. No "Import" pressed: the preamble is defined by the contents of the Preamble File (if specified).
2. If "Import" is pressed, the contents of the Preamble File are copied into the code box together with a "\begin{document}" line at the end.
3. If "\begin{document}" is present in the code box, the Preamble File is not used.

As a result, users can:
a. Work as usual, with a global Preamble File.
b. Import the Preamble File and tune it locally for a specific textext node.
c. Import the Preamble File and embed it as is to make the document more portable.

Just a  thought...

**Comment by jcwinkler** [comment]
Here some thoughts/ suggestions w.r.t. the topic of embedding the preamble in an Inkscape document. It is just a first draft to fix some ideas and might be a starting point to define a real concept about what has to be done.

**Definition:** “Local preamble file”: Preamble file residing on the machine Inkscape is running, i.e. the current situation. “Embedded preamble information”: Content of preamble file in some way included in Inkscape document.

**Current situation:** If someone moves an Inkscape file containing TexText nodes to another computer  compilation of the nodes might fail on this machine during re-editing if the preamble files on both machines do not match.

**Proposal:** Embed the preamble information in the Inkscape file or in the TexText nodes so one needs not to take care on the local preamble files on the different machines.

**Implementation issues:**

*Idea 1:* Embed the information on a per node basis. Advantage: If one copies the node from one Inkscape document to another one the preamble information is copied, too. Disadvantage: It is really uncomfortable to fiddle around with all the preamble information distributed on several nodes. I am against this solution.

*Idea 2:* Embed the information in the svg root of the document. Advantage: It is consistent to have the preamble information for the complete document. Disadvantage: If one copies a TexText node from one Inskcape document to another one the preamble information will not be copied as far as I do see. For the user this looks like if the information is lost. Nevertheless, I do favor this approach.

**User interface issues:**

Embedded preamble information needs to be editable in TexText window. User needs to decide whether to use embedded preamble information or local preamble file. User interface should not be overloaded. In most cases users are satisfied with current situation (local preamble file). 

*Idea 1:* Edit area for embedded preamble information is in a separate tab in the TexText window. Two radio buttons on front tab to define type of preamble information

*Idea 2:* Edit area for embedded preamble information AND local preamble file is in a separate tab in the TexText window. So front tab does only contain Texinput and layout, background tab is completely reserved for preamble stuff. (My favorite)

**Compatibility:**

It must be possible to open “old” TexText nodes created without this feature. In that case the local preamble file information option must be activated. Also, nodes created using the new approach should be editable with old TexText versions without this feature. This should not be too difficult to implement. New TexText versions could add an attribute “embedded-preamble=true/false” and everything is kept as is. So new TexText versions evaluate the new attribute and do the necessary actions while old ones just ignores them (with the danger that the nodes do not compile because the local preamble files does not contain what is in the embedded preamble information, but that’s in the responsibility of the user).

That’s it for now. You are invited to contribute.

**Comment by pacuserra** [comment]
I see the following dis/advantages of each preamble embedidng approach:

**Per node basis**
:heavy_check_mark: Copying one TexText node from document to document keeps preamble info.
:heavy_check_mark: Possibility to customize preamble for a particular given node.
:x: Changing preamble for all nodes requires editing each one.

**SVG root**
:heavy_check_mark: Uniform typesetting style for all nodes is ensured.
:x: Preamble info is lost when copying TexText nodes from document to document.
:x: Changing preamble requires recompilation (and reviewing) of all nodes in document!

I would vote for a mixed solution in the preamble tab:

1. Radio buttons for external/embedded preamble.
2. Common preamble section (for uniform typesetting, not ot be usually edited).
3. Node preamble section (for node tunning or full portability).

Just my thoughts...

**Comment by jcwinkler** [comment]
Thank you @pacuserra, that's this discussion is intended for.

I do see the benefits of your suggestion. It provides the most flexibility.

Question is if things do become too complicated for the average user. Difficult to estimate.

**Comment by pacuserra** [comment]
Your are welcome @jcwinkler. No user feedback can pay for such an excellent tool! :smiley:

I agree that adding too many options may distract the average user. Maybe the first radio button selector ('external'/'embedded') could be initiated to 'external' by default, and such a selection could then dim the rest of text boxes for 'common' and 'node' preambles...

**Comment by sizmailov** [comment]
I think we can manage to get best of two worlds using next scheme:

**Variables definition:**
`preamble_text` -- content of preamble
`preamble_hash` -- md5 hash of *preamble text*
`preamble_file` -- path to preamble file, just as now
`preamble_name` -- human-readable name
`textext_preambles` -- mapping of `preamble_hash` to  pair (`preamble_name`, `preamble_text`)

**Per node variables:** `preamble_hash`, `preamble_file`

**Per document variables:** `textext_preamble`

When new `preamble_text` entered/edited in GUI we 
 - store `preamble_text` at `~.config/inkscape/extensions/textext/preamble/<preamble_hash>.tex`
 - store mapping of  `preamble_hash` to `preamble_name` in `~.config/inkscape/extensions/textext/preambles.json` 

Thus `preamble_text` will not be repeated in `.svg` across the nodes, and will not produce errors for  node pasted from different document if we run textext on original document earlier on this computer before. 

**Disadvantages:**
 - `~.config/inkscape/extensions/textext/preamble/`might be populated with junk `.tex` files if you edit preamble too often.
 - if copy node from svg created on another computer you will have no matching preamble in  `~.config/inkscape/extensions/textext/preamble/`. Fortunately it's easy to detect and instruct user to run textext on foreign svg first. 


:warning: :warning: :warning: If we don't include full preamble text in each node we can run into situation when node has no corresponding preamble and cannot be re-compiled, so we get to the starting point. 

I think we should have two modes:

*Mode 1:* Root contains `textext_preambles`, nodes has `preamble_hash` and we store preambles locally as I described.

*Mode 2:* Full text of preamble for all nodes.

*Mode 1* is better than current *external preamble* mode, but has potential to break on copy-pasted nodes. 

*Mode 2* is rigid, but results to bloated `svg`

GUI should allow to convert all nodes of document at once. I don't think we need per-node granularity here, *preamble mode* should be global option.

**Comment by sizmailov** [comment]
Follow-up for @pacuserra list of dis/advantages:

**Mode 1 (see above)**
:heavy_check_mark:  Copying one TexText node from document to document keeps preamble info.
:heavy_check_mark: Possibility to customize preamble for a particular given node.

I can't say it's a disadvantage:
> :x: Changing preamble for all nodes requires editing each one. 

Node **must** be re-edited and recompiled, otherwise it will be inconsistent: `preamble+node_text != node_object`, but we can automate this action. 

**My comments about @pacuserra pro/cons**:
> :x: Uniform typesetting style for all nodes is ensured

No it's not. You still may have old-style nodes, which may use custom `preamble_file`.

> :x: Changing preamble requires recompilation (and reviewing) of all nodes in document!

Again, I think it's not a disadvantage, but rather the only right way of doing it, again, can be automated. 

@pacuserra Do you assume that there is just one preamble per svg? In principle it sounds reasonable: usually you want to have consistent TeX across a svg document. We may introduce a `single-preamble` mode which will be like `mode 1`, but limited to one preamble in document root.

**Comment by pacuserra** [comment]
@sizmailov, if I understood correctly, Mode 1 uses something like preamble **profiles**. The TeX code of each profile is stored in an external file, while the name/id mapping of all profiles is stored in SVG root, so you only need to store the profile id used by each node.

If that is the case, then I see:

:heavy_check_mark: Possibility to customize a **different** preamble profile for each node.
:heavy_check_mark: Changing the preamble of **all** nodes using the same profile only requires to edit that profile code.
:heavy_check_mark: Changing any preamble profile will recompile **only** those nodes using that profile.
:heavy_check_mark: Uniform typesetting style for a group (or all) nodes can be ensured by simply using a **single** profile.
:x: Preamble profiles codes are lost when **copying** TexText nodes from document to document.

...or do I miss anything here?

**Comment by sizmailov** [comment]
@pacuserra Your description is correct except one point: I propose to store TeX code of preamble at SVG root **and** in external file. 

If you copy a TexText node from one document to another you will be able to find TeX code of corresponding preamble in external file. (If you use this preamble on this computer before).

> Changing the preamble of all nodes using the same profile only requires to edit that profile code.

I have next scheme in my mind. If you edit current preamble profile you will obtain NEW preamble profile which supersedes old one, but old one is retained. Now you can trigger "upgrade" of all nodes: they will be recompiled with new profile, and old profile will be removed from SVG root (we need to store only minimal set of preambles). It's important to retain old file because otherwise you may unintentionally brake some other SVG files. Of course in GUI we can make "Save & Recompile all nodes" button to save some clicks. 

> Changing any preamble profile will recompile only those nodes using that profile.
> Uniform typesetting style for a group (or all) nodes can be ensured by simply using a single profile.

I guess a new GUI pane with list of all TexText nodes with previews, used preamble profiles, etc could be useful? It might include some batch operations on TexText nodes.

**Comment by sizmailov** [comment]
> So what's happening when trying to recompile a node on a different machine? TexText looks into svg-root if it does not find the file in the config dir?

Yes, TexText looks at svg root and config directory. Order doesn't really matter here because hash check insures the content will be same no matter where it was found. 

Config dir will be populated with new preambles (note: plural).

**Comment by pacuserra** [comment]
@sizmailov, I understand that when copying one TexText node from document to document, the preamble profile will be copied from the external file to the SVG root of the destination document (if not already there) for the purpose of portability, won't be?

**Comment by sizmailov** [comment]
@pacuserra Actually not. When you copy a node from one document to another SVG root gets no updates since TexText will not be triggered. SVG root will be updated if you run TexText plugin on any node of the document. 

This is the way to (accidentally) create a SVG document without preamble. Therefore the only way to protect users from that is including of preamble text to every node. Thus I think we should have two modes to fit every one needs.

**Comment by pacuserra** [comment]
@sizmailov, I see your point.

Then I would propose to store a copy of the preamble profile into every TexText node to ensure it is propagated when copying the node from document to document. In order to simplify the user interface, this could be done automatically, so user only sees a single mode of operation still based on profiles. 

In case the copied node is reedited in the destination document, then TexText could compare the hash of the preamble included in the node and add it to the SVG root / external files if not already existing.

What do you think?

**Comment by sizmailov** [comment]
What you described is more or less is how "preamble included in node" mode should work.

At first I thought it will heavily bloat resulting SVG with many TexText nodes, but actually preambles are usually small compared to SVG representation of TexText node. So we gain 100% portability in cost of a kilobyte or so. Now I think it is worth it and this mode should be default.

**Comment by pacuserra** [comment]
@sizmailov Fully agree, but still preserving your idea of preamble manager. In this way, user can manage a large number of TexText nodes with little effort because preambles are grouped in profiles...

**Comment by jcwinkler** [comment]
I am sorry, but I really see the danger that we tend to implement a tough and powerful mechanism handling all possible cases but which may only address the needs of a minor number of users in a few cases. I am not convinced that we need such a fine granular set of options.

What we should focus on is that the preamble content is available when moving the document to another machine. I am working with the extension for roughly 10 years now and hundreds of documents at our institute are based on it and I and my colleagues never run into a situation where we need a unique preamble for a certain node. 

Do you remember cases in your workflow where it has been indispensable to have a node specific preamble?

Lets wait with any implementations until we agreed on the features. Thanks!

**Comment by pacuserra** [comment]
@jcwinkler, if I had to choose one single feature, then I would still vote fro **portability** (as in my original post). Portability when I am sharing the SVG files with other people, and portability when I just want to reuse some particular TexText nodes from my old SVG files (i.e. copy & paste between documents).

Maybe we can come back to the basics and have just **two options**: the previous one with preamble referred to an external file **OR** the preamble code fully embedded in the TexText node. 

At the end, although we are using a formal typesetting language here, it is freedom/responsability of the designer to review/tune the compiled result for each TexText node (e.g. fitting a particular boundign box).

**Comment by pacuserra** [comment]
It looks like this enhancement request has languished...

Maybe my initial proposal could be reconsidered, as it requires a minimal modification and it is fully retrocompatible.

I am assuming here that the LaTeX code is in general composed of a concatenation of the contents of the preamble file + '\begin{document}' + contents of the textext code box + '\end{document}'. In that case, it could be as simple as:

**if** '\begin{document}' is present in the textext code,
   **then** LaTeX = preamble file + textext code + '\end{document}'
   **else** LaTeX = preamble file + '\begin{document}' + textext code + '\end{document}'

Hence, we would still enjoy the use of a global preamble, but with the possibility of adding extra packages to specific textext nodes. Or even, not use any global preamble at all (empty file) and include a full preamble in each node.

What do you think?

**Comment by sizmailov** [comment]
Yes, this feature has low priority, we are working on inkscape1.0 port right now

Quick note: @pacuserra Older versions of plugin will read preamble into text field, I think it's not acceptable behavior change

**Comment by tikuma-lsuhsc** [comment]
I just encountered an issue related to this: opening an svg file with TexText on another computer silently ignored the preamble file (which path is invalid on the second PC). 

I was able to recover the link by directly editing all preamble paths in the SVG file to relative paths. And TexText was able to find the correct preamble file (and promptly prepended the absolute path when modified as expected).

I like the idea of embedding the preamble, but it seems supporting relative path as commented a way earlier by @sizmailov seems to be the path of least resistance to allow preamble portability:

> @pacuserra Good call!
> 
> I had in mind adding extra preamble parameters such as "preamble path relative to svg", "preamble absolute path". It would be similar to what inkscape does for image links.

A UI checkbox with a minimal logic seems to be all we needed as the saved path can be checked for absolute/relative.

I'd be happy to work on this feature if such PR would be accepted.

P.S., I realize my suggestion may be a bit off topic. Let me know if this post should be posted as a new feature request.

**Comment by jcwinkler** [comment]
Hi @tikuma-lsuhsc Thanks for your suggestion. I also think the checkbox approach is reasonable. This can be implemented in TexText 2.0 which is currently re-written more or less from scratch. Please *do not work* with the actual develop source code. Most parts of TexText 2.0 are currently on my local machine until it has the stability to be pushed into develop. I will inform you when this is done and ready for testing and pull requests. 

I will prepare a mock-up of the new UI with this feature implemented during the next week.

**Comment by tikuma-lsuhsc** [comment]
@jcwinkler - Nice. Looking forward to the release of 2.0! Thanks.
